src/hotspot/share/gc/g1/g1EvacFailure.cpp

```diff
@@ -259,22 +260,20 @@ class RemoveSelfForwardPtrHRClosure: public HeapRegionClosure {
   }
 };
 
-G1ParRemoveSelfForwardPtrsTask::G1ParRemoveSelfForwardPtrsTask(G1RedirtyCardsQueueSet* rdcqs) :
+G1ParRemoveSelfForwardPtrsTask::G1ParRemoveSelfForwardPtrsTask(G1RedirtyCardsQueueSet* rdcqs,
+                                                               G1EvacFailureRegions* evac_failure_regions) :
   AbstractGangTask("G1 Remove Self-forwarding Pointers"),
   _g1h(G1CollectedHeap::heap()),
   _rdcqs(rdcqs),
   _hrclaimer(_g1h->workers()->active_workers()),
+  _evac_failure_regions(evac_failure_regions),
   _num_failed_regions(0) { }
 
 void G1ParRemoveSelfForwardPtrsTask::work(uint worker_id) {
   RemoveSelfForwardPtrHRClosure rsfp_cl(_rdcqs, worker_id, &_num_failed_regions);
 
-  // We need to check all collection set regions whether they need self forward
-  // removals, not only the last collection set increment. The reason is that
-  // reference processing (e.g. finalizers) can make it necessary to resurrect an
-  // otherwise unreachable object at the very end of the collection. That object
-  // might cause an evacuation failure in any region in the collection set.
-  _g1h->collection_set_par_iterate_all(&rsfp_cl, &_hrclaimer, worker_id);
+  // Iterates through only the regions recorded as evacuation failure.
```

Comment:
I would write something like:
```
// Iterate through all regions that failed evacuation during the entire collection.
```
But the code is kind of obvious to me now; the previous long-ish comment has only been there because to not over-optimize like has been done before. Your call.